Description: Use Debian sensible-browser
Forwarded: not-needed
Author: Frederic Peters <fpeters@debian.org>

--- a/epan/prefs.c
+++ b/epan/prefs.c
@@ -2676,7 +2676,7 @@
   prefs.gui_update_channel         = UPDATE_CHANNEL_STABLE;
   prefs.gui_update_interval        = 60*60*24; /* Seconds */
   /* This can be g_freed, so it must be g_mallocated. */
-  prefs.gui_webbrowser             = g_strdup(HTML_VIEWER " %s");
+  prefs.gui_webbrowser             = g_strdup("sensible-browser %s");
   /* This can be g_freed, so it must be g_mallocated. */
   prefs.gui_window_title           = g_strdup("");
   prefs.gui_start_title            = "The World's Most Popular Network Protocol Analyzer";
