diff --git a/epan/Makefile.am b/epan/Makefile.am
index 5eafa53..71a8ec7 100644
--- a/epan/Makefile.am
+++ b/epan/Makefile.am
@@ -56,7 +56,7 @@ ACLOCAL_AMFLAGS = `../aclocal-flags`
 noinst_LTLIBRARIES = libwireshark_generated.la libwireshark_asmopt.la
 lib_LTLIBRARIES = libwireshark.la
 # http://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
-libwireshark_la_LDFLAGS = -version-info 0:0:0 @LDFLAGS_SHAREDLIB@
+libwireshark_la_LDFLAGS = -version-info 3:0:0 @LDFLAGS_SHAREDLIB@
 
 include Makefile.common
 
diff --git a/wiretap/Makefile.am b/wiretap/Makefile.am
index fc430dc..c472a6d 100644
--- a/wiretap/Makefile.am
+++ b/wiretap/Makefile.am
@@ -30,7 +30,7 @@ include Makefile.common
 noinst_LTLIBRARIES = libwiretap_generated.la
 lib_LTLIBRARIES = libwiretap.la
 # http://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
-libwiretap_la_LDFLAGS = -version-info 0:0:0 @LDFLAGS_SHAREDLIB@
+libwiretap_la_LDFLAGS = -version-info 3:0:0 @LDFLAGS_SHAREDLIB@
 
 AM_NON_GENERATED_CFLAGS =-DWS_BUILD_DLL
 
diff --git a/wsutil/Makefile.am b/wsutil/Makefile.am
index 99e18ac..9fb1816 100644
--- a/wsutil/Makefile.am
+++ b/wsutil/Makefile.am
@@ -62,7 +62,7 @@ endif
 
 lib_LTLIBRARIES = libwsutil.la
 # http://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
-libwsutil_la_LDFLAGS = -version-info 0:0:0 @LDFLAGS_SHAREDLIB@
+libwsutil_la_LDFLAGS = -version-info 3:0:1 @LDFLAGS_SHAREDLIB@
 
 AM_CPPFLAGS = -I$(srcdir)/..
 
diff --git a/epan/CMakeLists.txt b/epan/CMakeLists.txt
index 2d022f3..43f591a 100644
--- a/epan/CMakeLists.txt
+++ b/epan/CMakeLists.txt
@@ -1550,11 +1550,11 @@ add_library(epan ${LINK_MODE_LIB}
 
 add_dependencies(epan svnversion)
 
-set(FULL_SO_VERSION "0.0.0")
+set(FULL_SO_VERSION "3.0.0")
 
 set_target_properties(epan PROPERTIES COMPILE_DEFINITIONS "WS_BUILD_DLL")
 set_target_properties(epan PROPERTIES LINK_FLAGS "${WS_LINK_FLAGS}")
-set_target_properties(epan PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 0)
+set_target_properties(epan PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 3)
 
 # discover and substitute list of include directories for ABI compatibility
 # checks
diff --git a/wiretap/CMakeLists.txt b/wiretap/CMakeLists.txt
index 5435482..ed99545 100644
--- a/wiretap/CMakeLists.txt
+++ b/wiretap/CMakeLists.txt
@@ -108,11 +108,11 @@ add_library(wiretap ${LINK_MODE_LIB}
 	${WIRETAP_FILES}
 )
 
-set(FULL_SO_VERSION "0.0.0")
+set(FULL_SO_VERSION "3.0.0")
 
 set_target_properties(wiretap PROPERTIES COMPILE_DEFINITIONS "WS_BUILD_DLL")
 set_target_properties(wiretap PROPERTIES LINK_FLAGS "${WS_LINK_FLAGS}")
-set_target_properties(wiretap PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 0)
+set_target_properties(wiretap PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 3)
 
 # discover and substitute list of include directories for ABI compatibility
 # checks
diff --git a/wsutil/CMakeLists.txt b/wsutil/CMakeLists.txt
index df689bf..e91f8cb 100644
--- a/wsutil/CMakeLists.txt
+++ b/wsutil/CMakeLists.txt
@@ -61,11 +61,11 @@ add_library(wsutil ${LINK_MODE_LIB}
   ${WSUTIL_FILES}
 )
 
-set(FULL_SO_VERSION "0.0.0")
+set(FULL_SO_VERSION "2.1.0")
 
 set_target_properties(wsutil PROPERTIES COMPILE_DEFINITIONS "WS_BUILD_DLL")
 set_target_properties(wsutil PROPERTIES LINK_FLAGS "${WS_LINK_FLAGS}")
-set_target_properties(wsutil PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 0)
+set_target_properties(wsutil PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 2)
 
 # discover and substitute list of include directories for ABI compatibility
 # checks
