diff --git a/epan/Makefile.am b/epan/Makefile.am
index 5eafa53..71a8ec7 100644
--- a/epan/Makefile.am
+++ b/epan/Makefile.am
@@ -56,7 +56,7 @@ ACLOCAL_AMFLAGS = `../aclocal-flags`
 noinst_LTLIBRARIES = libwireshark_generated.la libwireshark_asmopt.la
 lib_LTLIBRARIES = libwireshark.la
 # http://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
-libwireshark_la_LDFLAGS = -version-info 0:2:0 @LDFLAGS_SHAREDLIB@
+libwireshark_la_LDFLAGS = -version-info 3:0:0 @LDFLAGS_SHAREDLIB@
 
 include Makefile.common
 
diff --git a/wiretap/Makefile.am b/wiretap/Makefile.am
index fc430dc..c472a6d 100644
--- a/wiretap/Makefile.am
+++ b/wiretap/Makefile.am
@@ -30,7 +30,7 @@ include Makefile.common
 noinst_LTLIBRARIES = libwiretap_generated.la
 lib_LTLIBRARIES = libwiretap.la
 # http://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
-libwiretap_la_LDFLAGS = -version-info 0:2:0 @LDFLAGS_SHAREDLIB@
+libwiretap_la_LDFLAGS = -version-info 3:0:0 @LDFLAGS_SHAREDLIB@
 
 AM_NON_GENERATED_CFLAGS =-DWS_BUILD_DLL
 
diff --git a/wsutil/Makefile.am b/wsutil/Makefile.am
index 99e18ac..9fb1816 100644
--- a/wsutil/Makefile.am
+++ b/wsutil/Makefile.am
@@ -62,7 +62,7 @@ endif
 
 lib_LTLIBRARIES = libwsutil.la
 # http://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
-libwsutil_la_LDFLAGS = -version-info 0:0:0 @LDFLAGS_SHAREDLIB@
+libwsutil_la_LDFLAGS = -version-info 3:0:1 @LDFLAGS_SHAREDLIB@
 
 AM_CPPFLAGS = -I$(srcdir)/..
 
