diff --git a/epan/CMakeLists.txt b/epan/CMakeLists.txt
index 2d022f3..43f591a 100644
--- a/epan/CMakeLists.txt
+++ b/epan/CMakeLists.txt
@@ -1550,11 +1550,11 @@ add_library(epan ${LINK_MODE_LIB}
 
 add_dependencies(epan svnversion)
 
-set(FULL_SO_VERSION "0.0.0")
+set(FULL_SO_VERSION "3.0.0")
 
 set_target_properties(epan PROPERTIES COMPILE_DEFINITIONS "WS_BUILD_DLL")
 set_target_properties(epan PROPERTIES LINK_FLAGS "${WS_LINK_FLAGS}")
-set_target_properties(epan PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 0)
+set_target_properties(epan PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 3)
 
 # discover and substitute list of include directories for ABI compatibility
 # checks
diff --git a/wiretap/CMakeLists.txt b/wiretap/CMakeLists.txt
index 5435482..ed99545 100644
--- a/wiretap/CMakeLists.txt
+++ b/wiretap/CMakeLists.txt
@@ -108,11 +108,11 @@ add_library(wiretap ${LINK_MODE_LIB}
 	${WIRETAP_FILES}
 )
 
-set(FULL_SO_VERSION "0.0.0")
+set(FULL_SO_VERSION "3.0.0")
 
 set_target_properties(wiretap PROPERTIES COMPILE_DEFINITIONS "WS_BUILD_DLL")
 set_target_properties(wiretap PROPERTIES LINK_FLAGS "${WS_LINK_FLAGS}")
-set_target_properties(wiretap PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 0)
+set_target_properties(wiretap PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 3)
 
 # discover and substitute list of include directories for ABI compatibility
 # checks
diff --git a/wsutil/CMakeLists.txt b/wsutil/CMakeLists.txt
index df689bf..e91f8cb 100644
--- a/wsutil/CMakeLists.txt
+++ b/wsutil/CMakeLists.txt
@@ -61,11 +61,11 @@ add_library(wsutil ${LINK_MODE_LIB}
   ${WSUTIL_FILES}
 )
 
-set(FULL_SO_VERSION "0.0.0")
+set(FULL_SO_VERSION "3.0.0")
 
 set_target_properties(wsutil PROPERTIES COMPILE_DEFINITIONS "WS_BUILD_DLL")
 set_target_properties(wsutil PROPERTIES LINK_FLAGS "${WS_LINK_FLAGS}")
-set_target_properties(wsutil PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 0)
+set_target_properties(wsutil PROPERTIES VERSION ${FULL_SO_VERSION} SOVERSION 3)
 
 # discover and substitute list of include directories for ABI compatibility
 # checks
