Description: Upstream-derived patch to attempt to fix FTBFS
Origin:      http://anonsvn.wireshark.org/viewvc/trunk/configure.ac?r1=54337&r2=54336&pathrev=54337&view=patch

Index: b/configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -1390,7 +1390,13 @@
 
 	CPPFLAGS="-DGDK_PIXBUF_DISABLE_DEPRECATED $CPPFLAGS"
 	CPPFLAGS="-DGDK_DISABLE_DEPRECATED $CPPFLAGS"
-	CPPFLAGS="-DGTK_DISABLE_DEPRECATED $CPPFLAGS"
+        if test \( $gtk_config_major_version -eq 3 -a $gtk_config_minor_version -ge 10 \) ; then
+                ## Allow use of deprecated & disable deprecated warnings if Gtk >= 3.10;
+                ##  The deprecations in Gtk 3.10 will not be fixed ...
+                CPPFLAGS="-DGDK_DISABLE_DEPRECATION_WARNINGS $CPPFLAGS"
+        else
+                CPPFLAGS="-DGTK_DISABLE_DEPRECATED $CPPFLAGS"
+        fi
 	CPPFLAGS="-DGTK_DISABLE_SINGLE_INCLUDES $CPPFLAGS"
 	if test ! \( $gtk_config_major_version -eq 2 -a $gtk_config_minor_version -lt 20 \) ; then
 		# Enable GSEAL when building with GTK > 2.20
