Description: Replace g_memmove() with memmove() to fix FTBFS
Author:      Thomas Ward <teward@ubuntu.com>

Index: b/ui/export_object_smb.c
===================================================================
--- a/ui/export_object_smb.c
+++ b/ui/export_object_smb.c
@@ -254,7 +254,7 @@
 	/* ...then, put the chunk of the file in the right place */
 	if (!file->is_out_of_memory) {
 		dest_memory_addr = entry->payload_data + chunk_offset;
-		g_memmove(dest_memory_addr, eo_info->payload_data, eo_info->payload_len);
+		memmove(dest_memory_addr, eo_info->payload_data, eo_info->payload_len);
 	}
 }
 
Index: b/epan/dissectors/packet-ssl-utils.c
===================================================================
--- a/epan/dissectors/packet-ssl-utils.c
+++ b/epan/dissectors/packet-ssl-utils.c
@@ -1633,7 +1633,7 @@
     ssl_debug_printf("pcry_private_decrypt: stripping %d bytes, decr_len %" G_GSIZE_MODIFIER "u\n",
         rc, decr_len);
     ssl_print_data("decrypted_unstrip_pre_master", decr_data_ptr, decr_len);
-    g_memmove(decr_data_ptr, &decr_data_ptr[rc], decr_len - rc);
+    memmove(decr_data_ptr, &decr_data_ptr[rc], decr_len - rc);
     decr_len -= rc;
 
 out:
@@ -1670,7 +1670,7 @@
     ssl_debug_printf("pcry_private_decrypt: stripping %d bytes, decr_len %d\n",
         rc, decr_len);
     ssl_print_data("decrypted_unstrip_pre_master", decr_data_ptr, decr_len);
-    g_memmove(decr_data_ptr, &decr_data_ptr[rc], decr_len - rc);
+    memmove(decr_data_ptr, &decr_data_ptr[rc], decr_len - rc);
     decr_len -= rc;
 #endif /* SSL_FAST */
     gcry_mpi_release(text);
